File: cpp/submod/LogDeterminant.cpp
// Implementation of Algorithm 1 in
// https://proceedings.neurips.cc/paper/2018/file/dbbf603ff0e99629dda5d75b6f75f966-Paper.pdf

#include <algorithm>
#include <cmath>
#include <iostream>
#include <iterator>
#include <map>
#include <set>
#include <string>
#include <vector>

#include "../utils/helper.h"
#include "LogDeterminant.h"

double dotProduct(const std::vector<double> &x, const std::vector<double> &y) {
    double result = 0;
    for (int i = 0; i < x.size(); i++) {
        result += x[i] * y[i];
    }
    return result;
}

LogDeterminant::LogDeterminant() {}

// Constructor for dense mode
LogDeterminant::LogDeterminant(
    ll n_, std::vector<std::vector<float>> const &denseKernel_, bool partial_,
    std::unordered_set<ll> const &ground_, double lambda_)
    : n(n_),
      mode(dense),
      denseKernel(denseKernel_),
      partial(partial_),
      lambda(lambda_) {
    if (partial == true) {
        // ground set will now be the subset provided
        effectiveGroundSet = ground_;
    } else {
        // create groundSet with items 0 to n-1
        effectiveGroundSet.reserve(n);
        for (ll i = 0; i < n; ++i) {
            effectiveGroundSet.insert(i);  // each insert takes O(1) time
        }
    }
    numEffectiveGroundset = effectiveGroundSet.size();
    memoizedC = std::vector<std::vector<double>>(numEffectiveGroundset, std::vector<double>());
    prevDetVal = 0;
    memoizedD.clear();
    prevItem = -1;
    if (partial == true) {
        ll ind = 0;
        for (auto it : effectiveGroundSet) {
            //std::cout << it << "-->" << ind << "\n";
            originalToPartialIndexMap[it] = ind;
            ind += 1;
            memoizedD.push_back(sqrt(denseKernel[it][it] + lambda));
        }
    } else {
        for (ll i = 0; i < n; i++) {
            memoizedD.push_back(sqrt(denseKernel[i][i] + lambda));
        }
    }
    // std::cout << "LogDet Constructor: MemoizedD: [";
    // for(auto temp: memoizedD) {
    //     std::cout << temp << ", ";
    // }
    // std::cout << "]\n";
}

// Constructor for sparse mode
LogDeterminant::LogDeterminant(ll n_, std::vector<float> const &arr_val,
                               std::vector<ll> const &arr_count,
                               std::vector<ll> const &arr_col, double lambda_)
    : n(n_), mode(sparse), partial(false), lambda(lambda_) {
    if (arr_val.size() == 0 || arr_count.size() == 0 || arr_col.size() == 0) {
        throw "Error: Empty/Corrupt sparse similarity kernel";
    }
    sparseKernel = SparseSim(arr_val, arr_count, arr_col);
    effectiveGroundSet.reserve(n);
    for (ll i = 0; i < n; ++i) {
        effectiveGroundSet.insert(i);  // each insert takes O(1) time
    }
    numEffectiveGroundset = effectiveGroundSet.size();
    memoizedC = std::vector<std::vector<double>>(n, std::vector<double>());
    memoizedD.clear();
    prevDetVal = 0;
    for (ll i = 0; i < n; i++) {
        memoizedD.push_back(sqrt(sparseKernel.get_val(i, i) + lambda));
    }
    prevItem = -1;
    // std::cout << "MemoizedD: [";
    // for(auto temp: memoizedD) {
    //     std::cout << temp << ", ";
    // }
    // std::cout << "]\n";
    // std::cout << "Previous detVal = " << prevDetVal << "\n";
}

LogDeterminant::LogDeterminant(const LogDeterminant& f)
    : n(f.n),
      mode(f.mode),
      denseKernel(f.denseKernel),
      sparseKernel(f.sparseKernel),
      partial(f.partial),
      lambda(f.lambda), effectiveGroundSet(f.effectiveGroundSet), numEffectiveGroundset(f.numEffectiveGroundset), memoizedC(f.memoizedC), memoizedD(f.memoizedD), prevDetVal(f.prevDetVal), prevItem(f.prevItem), originalToPartialIndexMap(f.originalToPartialIndexMap) {
}

LogDeterminant* LogDeterminant::clone() {
    return new LogDeterminant(*this);
}

double LogDeterminant::evaluate(std::unordered_set<ll> const &X) {
    // since computeDeterminant doesnt work for bigger ground sets, calculating
    // eval through evalFast however, since eval should not affect the current
    // state of preCompute, restoring it to the value before
    // std::cout << "LogDet eval called\n";
    std::vector<std::vector<double>> currMemoizedC = memoizedC;
    std::vector<double> currMemoizedD = memoizedD;
    int currprevItem = prevItem;
    double currprevDetVal = prevDetVal;
    setMemoization(X);
    // std::cout << "Memoization set\n";
    // std::cout << "MemoizedD: [";
    // for(auto temp: memoizedD) {
    //     std::cout << temp << ", ";
    // }
    // std::cout << "]\n";
    double result = evaluateWithMemoization(X);
    // restore Memoized to what they were
    memoizedC = currMemoizedC;
    memoizedD = currMemoizedD;
    prevItem = currprevItem;
    prevDetVal = currprevDetVal;
    return result;
}

double LogDeterminant::evaluateWithMemoization(
    std::unordered_set<ll> const &X) {
    return prevDetVal;
}

double LogDeterminant::marginalGain(std::unordered_set<ll> const &X, ll item) {
    std::vector<std::vector<double>> currMemoizedC = memoizedC;
    std::vector<double> currMemoizedD = memoizedD;
    int currprevItem = prevItem;
    double currprevDetVal = prevDetVal;
    setMemoization(X);
    double result = marginalGainWithMemoization(X, item);
    // restore Memoized to what they were
    memoizedC = currMemoizedC;
    memoizedD = currMemoizedD;
    prevItem = currprevItem;
    prevDetVal = currprevDetVal;
    return result;
}

double LogDeterminant::marginalGainWithMemoization(
    std::unordered_set<ll> const &X, ll item, bool enableChecks) {
    // std::cout << "LogDet's marginalGainWithMemoization called with X={";
    // for(auto temp: X) {
    //     std::cout << temp << ", ";
    // }
    // std::cout << "} and item=" << item << "\n";
    //this assumes that prevItem was the previous best, for example, when called in context of maximization
    std::unordered_set<ll> effectiveX;
    double gain = 0;

    if (partial == true) {
        // effectiveX = intersect(X, effectiveGroundSet)
        // std::set_intersection(X.begin(), X.end(), effectiveGroundSet.begin(),
        //                       effectiveGroundSet.end(),
        //                       std::inserter(effectiveX, effectiveX.begin()));
        effectiveX = set_intersection(X, effectiveGroundSet);
    } else {
        effectiveX = X;
    }

    if (enableChecks && effectiveX.find(item) != effectiveX.end()) {
        return 0;
    }
    if (partial && effectiveGroundSet.find(item)==effectiveGroundSet.end()) {
        return 0;
    }
    ll itemIndex = (partial)?originalToPartialIndexMap[item]:item;

    if (mode == dense) {
        if (effectiveX.size() == 0) {
            gain = log(memoizedD[itemIndex] * memoizedD[itemIndex]);
            // std::cout << "Gain on empty set = " << gain << "\n";
        } else if (effectiveX.size() == 1) {
            ll prevItemIndex = (partial)?originalToPartialIndexMap[prevItem]:prevItem;
            double e = denseKernel[prevItem][item] / memoizedD[prevItemIndex];
            gain = log(memoizedD[itemIndex] * memoizedD[itemIndex] - e * e);
            // std::cout << "Gain on set of size 1 = " << gain << "\n";
        } else {
            ll prevItemIndex = (partial)?originalToPartialIndexMap[prevItem]:prevItem;
            // std::cout << "prevItemIndex = " << prevItemIndex << "\n";
            double e = (denseKernel[prevItem][item] -
                        dotProduct(memoizedC[prevItemIndex], memoizedC[itemIndex])) /
                       memoizedD[prevItemIndex];
            // std::cout << "e = " << e << "\n";
            // std::cout << "memoizedD[itemIndex] * memoizedD[itemIndex] = " << memoizedD[itemIndex] * memoizedD[itemIndex] << "\n";
            gain = log(memoizedD[itemIndex] * memoizedD[itemIndex] - e * e);
            // std::cout << "gain (3rd condition) = " << gain << "\n";
        }
    } else if (mode == sparse) {
        if (effectiveX.size() == 0) {
            gain = log(memoizedD[itemIndex] * memoizedD[itemIndex]);
        } else if (effectiveX.size() == 1) {
            ll prevItemIndex = (partial)?originalToPartialIndexMap[prevItem]:prevItem;
            double e =
                sparseKernel.get_val(prevItem, item) / memoizedD[prevItemIndex];
            gain = log(memoizedD[itemIndex] * memoizedD[itemIndex] - e * e);
        } else {
            ll prevItemIndex = (partial)?originalToPartialIndexMap[prevItem]:prevItem;
            // std::cout << "prevItemIndex = " << prevItemIndex << "; ";
            // std::cout << "sparseKernel.get_val(prevItem, item) = " << sparseKernel.get_val(prevItem, item) << "; ";
            // std::cout << "memoizedD[prevItemIndex] = " << memoizedD[prevItemIndex] <<"; ";
            double e = (sparseKernel.get_val(prevItem, item) -
                        dotProduct(memoizedC[prevItemIndex], memoizedC[itemIndex])) /
                       memoizedD[prevItemIndex];
            // std::cout << "memoizedD[itemIndex] = " << memoizedD[itemIndex] << "; ";
            // std::cout << "e = " << e <<"\n";
            gain = log(memoizedD[itemIndex] * memoizedD[itemIndex] - e * e);
        }
    } else {
        throw "Error: Only dense and sparse mode supported";
    }

    return gain;
}

void LogDeterminant::updateMemoization(std::unordered_set<ll> const &X,
                                       ll item) {
    //TODO: this assumes that prevItem was the previous best, for example when invoked in context of maximization
    // std::cout << "LogDeterminant updateMemoization\n";
    std::unordered_set<ll> effectiveX;

    if (partial == true) {
        // effectiveX = intersect(X, effectiveGroundSet)
        // std::set_intersection(X.begin(), X.end(), effectiveGroundSet.begin(),
        // effectiveGroundSet.end(), std::inserter(effectiveX,
        // effectiveX.begin()));
        effectiveX = set_intersection(X, effectiveGroundSet);
    } else {
        effectiveX = X;
    }
    if (effectiveX.find(item) != effectiveX.end()) {
        return;
    }
    if (effectiveGroundSet.find(item)==effectiveGroundSet.end()) {
        return;
    }

    prevDetVal += marginalGainWithMemoization(X, item);
    //std::cout << "New previous detVal = " << prevDetVal << "\n";

    if (effectiveX.size() == 0) {
        // nothing needs to be done
        // memoizedD is already set appropriately
    } else {
        ll prevItemIndex = (partial)?originalToPartialIndexMap[prevItem]:prevItem;
        double prevDValue = memoizedD[prevItemIndex];
        // std::cout << "prevItemIndex = " << prevItemIndex << "; ";
        // std::cout << "prevDValue = " << prevDValue << "; ";
        if (mode == dense) {
            for(auto i: effectiveGroundSet) {
                ll iIndex = (partial)?originalToPartialIndexMap[i]:i;
                if (effectiveX.find(i) != effectiveX.end()) {
                    // std::cout << "Skipping i of " << i << " as it is already
                    // in sset" << std::endl;
                    continue;
                }
                double e = 0;
                if (effectiveX.size() == 1) {
                    // e = kernel[prevItem][i] / prevDValue;
                    e = denseKernel[prevItem][i] / memoizedD[prevItemIndex];
                    // vector<double> currvecC = vector<double>();
                    // currvecC.push_back(e);
                    // preComputeC.push_back(currvecC);
                    memoizedC[iIndex].push_back(e);
                } else {
                    // e = (denseKernel[prevItem][i] -
                    //      datk::innerProduct(memoizedC[prevItem],
                    //                         memoizedC[i])) /
                    //     prevDValue;
                    e = (denseKernel[prevItem][i] -
                         dotProduct(memoizedC[prevItemIndex], memoizedC[iIndex])) /
                        memoizedD[prevItemIndex];
                    memoizedC[iIndex].push_back(e);
                }
                // std::cout << "e = " << e << "\n";
                // std::cout << "memoizedD[iIndex] * memoizedD[iIndex] = " << memoizedD[iIndex] * memoizedD[iIndex] << "\n";
                memoizedD[iIndex] = sqrt(memoizedD[iIndex] * memoizedD[iIndex] - e * e);
            }

        } else if (mode == sparse) {
            //std::cout << "Updating memoizedC and memoizedD\n";
            for (ll i = 0; i < n; i++) {
                //std::cout << i << ": ";
                if (effectiveX.find(i) != effectiveX.end()) {
                    //std::cout << "skipping\n";
                    continue;
                }
                double e = 0;
                if (effectiveX.size() == 1) {
                    // e = kernel[prevItem][i] / prevDValue;
                    e = sparseKernel.get_val(prevItem, i) / memoizedD[prevItem];
                    // vector<double> currvecC = vector<double>();
                    // currvecC.push_back(e);
                    // preComputeC.push_back(currvecC);
                    memoizedC[i].push_back(e);
                } else {
                    // e = (sparseKernel.get_val(prevItem, i) -
                    //      datk::innerProduct(memoizedC[prevItem],
                    //                         memoizedC[i])) /
                    //     prevDValue;
                    e = (sparseKernel.get_val(prevItem, i) -
                         dotProduct(memoizedC[prevItem], memoizedC[i])) /
                        memoizedD[prevItem];
                    memoizedC[i].push_back(e);
                }
                memoizedD[i] = sqrt(memoizedD[i] * memoizedD[i] - e * e);
                //std::cout << "e = " << e << " memoizedD[i] = " << memoizedD[i] << "\n";
                // if(std::isnan(memoizedD[i])) {
                //     std::cout << "memoizedD[i] has become NAN!!";
                //     exit(1);
                // }
            }

        } else {
            throw "Error: Only dense and sparse mode supported";
        }
    }
    prevItem = item;
}

std::unordered_set<ll> LogDeterminant::getEffectiveGroundSet() {
    return effectiveGroundSet;
}

// std::vector<std::pair<ll, double>> LogDeterminant::maximize(
//     std::string optimizer, ll budget, bool stopIfZeroGain = false,
//     bool stopIfNegativeGain = false, float epsilon = 0.1,
//     bool verbose = false, bool showProgress=true) {
//     // std::cout << "LogDeterminant maximize\n";
//     if (optimizer == "NaiveGreedy") {
//         return NaiveGreedyOptimizer().maximize(*this, budget, stopIfZeroGain,
//                                                stopIfNegativeGain, verbose, showProgress);
//     } else if (optimizer == "LazyGreedy") {
//         return LazyGreedyOptimizer().maximize(*this, budget, stopIfZeroGain,
//                                               stopIfNegativeGain, verbose, showProgress);
//     } else if (optimizer == "StochasticGreedy") {
//         return StochasticGreedyOptimizer().maximize(
//             *this, budget, stopIfZeroGain, stopIfNegativeGain, epsilon,
//             verbose, showProgress);
//     } else if (optimizer == "LazierThanLazyGreedy") {
//         return LazierThanLazyGreedyOptimizer().maximize(
//             *this, budget, stopIfZeroGain, stopIfNegativeGain, epsilon, verbose, showProgress);
//     } else {
//         throw "Invalid optimizer";
//     }
// }

void LogDeterminant::cluster_init(
    ll n_, std::vector<std::vector<float>> const &denseKernel_,
    std::unordered_set<ll> const &ground_, bool partial, float lambda) {
    // std::cout << "LogDeterminant clusterInit\n";
    *this = LogDeterminant(n_, denseKernel_, partial, ground_, lambda);
}

void LogDeterminant::clearMemoization() {
    // std::cout << "LogDet clearMemoization()\n";
    memoizedC.clear();
    memoizedC = std::vector<std::vector<double>>(numEffectiveGroundset, std::vector<double>());
    prevDetVal = 0;
    prevItem = -1;
    if (mode == dense) {
        if (partial == true) {
            for (auto it : effectiveGroundSet) {
                ll index = originalToPartialIndexMap[it];
                memoizedD[index] = sqrt(denseKernel[it][it] + lambda);
            }
        } else {
            for (ll i = 0; i < n; i++) {
                memoizedD[i] = sqrt(denseKernel[i][i] + lambda);
            }
        }
    } else if (mode == sparse) {
        for (ll i = 0; i < n; i++) {
            memoizedD[i] = sqrt(sparseKernel.get_val(i, i) + lambda);
        }
    } else {
        throw "Error: Only dense and sparse mode supported";
    }
    //std::cout << "After clear memoization: \n";
    // std::cout << "MemoizedC: [";
    // for(auto temp: memoizedC) {
    //     std::cout << temp << ", ";
    // }
    // std::cout << "]\n";
    // std::cout << "MemoizedD: [";
    // for(auto temp: memoizedD) {
    //     std::cout << temp << ", ";
    // }
    // std::cout << "]\n";
    // std::cout << "Previous detVal = " << prevDetVal << "\n";
}

void LogDeterminant::setMemoization(std::unordered_set<ll> const &X) {

    //TODO: this implementation is probably wrong. I guess for updateMemoization to work correctly, the items should be added in context of greedy maximization and not arbitrarily
    // std::cout << "LogDeterminant setMemoization\n";
    clearMemoization();
    std::unordered_set<ll> temp;
    // ll best_id;
	// double best_val;
    // for(int count = 0; count < X.size(); count++) {
    //     best_id = -1;
    //     best_val = -1 * std::numeric_limits<double>::max();
    //     for (auto i: X) {
    //         if (temp.find(i) != temp.end()) { 
    //             //if this datapoint has already been included in temp, skip it
    //             continue;
    //         }
    //         double gain = marginalGainWithMemoization(temp, i);
    //         if (std::isnan(gain)) {
    //             throw "Gain turned out to be NAN";
    //         }
    //         if (gain > best_val) {
    //             best_id = i;
    //             best_val = gain;
    //         }
    //     }
    //     updateMemoization(temp, best_id);
    //     temp.insert(best_id);
    // }
    
    for (auto elem : X) {
        // std::cout << "Memoization for adding " << elem << " to {";
        // for (auto tmp: temp) {
        //     std::cout << tmp <<", ";
        // }
        // std::cout << "}\n";
        updateMemoization(temp, elem);
        temp.insert(elem);
    }
}

File: pytorch/submod/LogDeterminant.py
import math
from collections import defaultdict
import scipy
from helper import *
from ..SetFunction import SetFunction

class LogDeterminantFunction(SetFunction):

    def dot_product(self, x, y):
        return sum(xi * yi for xi, yi in zip(x, y))


    def __init__(self, n, mode, lambdaVal, arr_val=None, arr_count=None, arr_col=None, dense_kernel=None, partial=None,
                  sijs=None, data=None, metric="cosine", num_neighbors=None, memoizedC = None, memoizedD = None, data_master = None):
        self.n = n
        self.mode = mode
        self.metric = metric
        self.sijs = sijs
        self.data = data
        self.num_neighbors = num_neighbors
        self.lambdaVal = lambdaVal
        self.sijs = None
        self.content = None
        self.effective_ground = None
        self.partial = partial
        self.effective_ground_set = set(range(n))
        self.memoizedC = memoizedC
        self.memoizedD = memoizedD
        self.data_master = data_master
        self.dense_kernel = dense_kernel

        if self.n <= 0:
          raise Exception("ERROR: Number of elements in ground set must be positive")

        if self.mode not in ['dense', 'sparse', 'clustered']:
          raise Exception("ERROR: Incorrect mode. Must be one of 'dense', 'sparse' or 'clustered'")

        if self.metric not in ['euclidean', 'cosine']:
        	raise Exception("ERROR: Unsupported metric. Must be 'euclidean' or 'cosine'")
        if type(self.sijs) != type(None): # User has provided similarity kernel
          if type(self.sijs) == scipy.sparse.csr.csr_matrix:
            if num_neighbors is None or num_neighbors <= 0:
              raise Exception("ERROR: Positive num_neighbors must be provided for given sparse kernel")
            if mode != "sparse":
              raise Exception("ERROR: Sparse kernel provided, but mode is not sparse")
          elif type(self.sijs) == np.ndarray:
            if mode != "dense":
              raise Exception("ERROR: Dense kernel provided, but mode is not dense")
          else:
            raise Exception("Invalid kernel provided")
          #TODO: is the below dimensionality check valid for both dense and sparse kernels?
          if np.shape(self.sijs)[0]!=self.n or np.shape(self.sijs)[1]!=self.n:
            raise Exception("ERROR: Inconsistentcy between n and dimensionality of given similarity kernel")
          if type(self.data) != type(None):
            print("WARNING: similarity kernel found. Provided data matrix will be ignored.")
        else: #similarity kernel has not been provided
          if type(self.data) != type(None):
            if np.shape(self.data)[0]!=self.n:
              raise Exception("ERROR: Inconsistentcy between n and no of examples in the given data matrix")

            if self.mode == "dense":
              if self.num_neighbors  is not None:
                raise Exception("num_neighbors wrongly provided for dense mode")
              self.num_neighbors = np.shape(self.data)[0] #Using all data as num_neighbors in case of dense mode
            self.content = np.array(create_kernel( X = self.data.tolist(), metric = self.metric, mode = self.mode, num_neigh = self.num_neighbors))
            val = self.content[0]
            row = list(self.content[1].astype(int))
            col = list(self.content[2].astype(int))
            if self.mode=="dense":
              self.sijs = np.zeros((n,n))
              self.sijs[row,col] = val
            if self.mode=="sparse":
              self.sijs = sparse.csr_matrix((val, (row, col)), [n,n])
          else:
            raise Exception("ERROR: Neither ground set data matrix nor similarity kernel provided")


        #Breaking similarity matrix to simpler native data structures for implicit pybind11 binding
        if self.mode=="dense":
          self.sijs = self.sijs.tolist() #break numpy ndarray to native list of list datastructure

          if type(self.sijs[0])==int or type(self.sijs[0])==float: #Its critical that we pass a list of list to pybind11
                                          #This condition ensures the same in case of a 1D numpy array (for 1x1 sim matrix)
            l=[]
            l.append(self.sijs)
            self.sijs=l

        self.effective_ground = self.get_effective_ground_set()
        if self.mode == 'dense':
          if self.dense_kernel == None:
             self.dense_kernel = create_kernel_NS(X_ground = self.data, X_master = self.data, metric = self.metric)
          if self.partial:
            self.effectiveGroundSet = self.data
          else:
            self.effectiveGroundSet = set(range(n))
            self.numEffectiveGroundset = len(self.effectiveGroundSet)
            self.memoizedC = [[] for _ in range(self.numEffectiveGroundset)]
            self.prevDetVal = 0
            self.memoizedD = []
            self.prevItem = -1

            if self.partial:
                ind = 0
                for it in self.effectiveGroundSet:
                    self.originalToPartialIndexMap[it] = ind
                    ind += 1
                    self.memoizedD.append(np.sqrt(self.dense_kernel[it][it] + self.lambdaVal))
            else:
                for i in range(self.n):
                    self.memoizedD.append(np.sqrt(self.dense_kernel[i][i] + self.lambdaVal))

        elif arr_val is not None and arr_count is not None and arr_col is not None:
            self.n = n
            self.mode = 'sparse'
            self.lambdaVal = lambdaVal
            self.sparseKernel = SparseSim(arr_val, arr_count, arr_col)
            self.effectiveGroundSet = set(range(n_))
            self.numEffectiveGroundset = len(self.effectiveGroundSet)
            self.memoizedC = [[] for _ in range(n_)]
            self.memoizedD = []
            self.prevDetVal = 0
            self.prevItem = -1

            for i in range(self.n):
                self.memoizedD.append(np.sqrt(self.sparseKernel.get_val(i, i) + self.lambdaVal))

        else:
            raise ValueError("Invalid constructor arguments. Please provide either denseKernel or sparse kernel data.")

    def evaluate(self, X):
        currMemoizedC = self.memoizedC.copy()
        currMemoizedD = self.memoizedD.copy()
        currprevItem = self.prevItem
        currprevDetVal = self.prevDetVal
        self.setMemoization(X)
        result = self.evaluate_with_memoization(X)
        self.memoizedC = currMemoizedC
        self.memoizedD = currMemoizedD
        self.prevItem = currprevItem
        self.prevDetVal = currprevDetVal
        return result

    def evaluate_with_memoization(self, X):
        return self.prevDetVal

    def marginal_gain(self, X, item):
        currMemoizedC = self.memoizedC.copy()
        currMemoizedD = self.memoizedD.copy()
        currprevItem = self.prevItem
        currprevDetVal = self.prevDetVal
        self.set_memoization(X)
        result = self.marginal_gain_with_memoization(X, item)
        self.memoizedC = currMemoizedC
        self.memoizedD = currMemoizedD
        self.prevItem = currprevItem
        self.prevDetVal = currprevDetVal
        return result

    def marginal_gain_with_memoization(self, X, item, enableChecks=True):
        effectiveX = X.intersection(self.effective_ground_set) if self.partial else X
        gain = 0

        if enableChecks and item in effectiveX:
            return 0

        if self.partial and item not in self.effective_ground_set:
            return 0

        itemIndex = self.originalToPartialIndexMap[item] if self.partial else item

        if self.mode == "dense":
            if len(effectiveX) == 0:
                gain = math.log(self.memoizedD[itemIndex] * self.memoizedD[itemIndex])
            elif len(effectiveX) == 1:
                prevItemIndex = self.originalToPartialIndexMap[self.prevItem] if self.partial else self.prevItem
                e = self.dense_kernel[self.prevItem][item] / self.memoizedD[prevItemIndex]
                gain = math.log(math.fabs(self.memoizedD[itemIndex] * self.memoizedD[itemIndex] - e * e))
            else:
                prevItemIndex = self.originalToPartialIndexMap[self.prevItem] if self.partial else self.prevItem
                e = (self.dense_kernel[self.prevItem][item] -
                     self.dot_product(self.memoizedC[prevItemIndex], self.memoizedC[itemIndex])) / self.memoizedD[prevItemIndex]
                gain = math.log(math.fabs(self.memoizedD[itemIndex] * self.memoizedD[itemIndex] - e * e))
        elif self.mode == "sparse":
            if len(effectiveX) == 0:
                gain = math.log(math.fabs(self.memoizedD[itemIndex] * self.memoizedD[itemIndex]))
            elif len(effectiveX) == 1:
                prevItemIndex = self.originalToPartialIndexMap[self.prevItem] if self.partial else self.prevItem
                e = self.sparseKernel.get_val(self.prevItem, item) / self.memoizedD[prevItemIndex]
                gain = math.log(math.fabs(self.memoizedD[itemIndex] * self.memoizedD[itemIndex] - e * e))
            else:
                prevItemIndex = self.originalToPartialIndexMap[self.prevItem] if self.partial else self.prevItem
                e = (self.sparseKernel.get_val(self.prevItem, item) -
                     self.dot_product(self.memoizedC[prevItemIndex], self.memoizedC[itemIndex])) / self.memoizedD[prevItemIndex]
                gain = math.log(math.fabs(self.memoizedD[itemIndex] * self.memoizedD[itemIndex] - e * e))
        else:
            raise ValueError("Only dense and sparse mode supported")

        return gain

    def update_memoization(self, X, item):
        effectiveX = X.intersection(self.effective_ground_set) if self.partial else X

        if item in effectiveX:
            return

        if item not in self.effective_ground_set:
            return

        self.prevDetVal += self.marginal_gain_with_memoization(X, item)

        if len(effectiveX) == 0:
            pass
        else:
            prevItemIndex = self.originalToPartialIndexMap[self.prevItem] if self.partial else self.prevItem
            prevDValue = self.memoizedD[prevItemIndex]

            for i in self.effectiveGroundSet:
                iIndex = self.originalToPartialIndexMap[i] if self.partial else i

                if i in effectiveX:
                    continue

                e = 0
                if len(effectiveX) == 1:
                    e = self.dense_kernel[self.prevItem][i] / prevDValue
                    self.memoizedC[iIndex].append(e)
                else:
                    e = (self.dense_kernel[self.prevItem][i] -
                         self.dot_product(self.memoizedC[prevItemIndex], self.memoizedC[iIndex])) / prevDValue
                    self.memoizedC[iIndex].append(e)

                self.memoizedD[iIndex] = math.sqrt(math.fabs(self.memoizedD[iIndex] * self.memoizedD[iIndex] - e * e))

        self.prevItem = item

    def get_effective_ground_set(self):
        return self.effective_ground_set

    def clear_memoization(self):
        self.memoizedC.clear()
        self.memoizedC = defaultdict(list)
        self.prevDetVal = 0
        self.prevItem = -1

        if self.mode == "dense":
            if self.partial:
                for it in self.effective_ground_set:
                    index = self.originalTo
